psftools/man/txt2psf.1
.\" -*- nroff -*-
.\"
.\" txt2psf.1: psf2txt man page
.\" Copyright (c) 2005, 2007 John Elliott
.\"
.\"
.\"
.\" psftools: Manipulate console fonts in the .PSF format
.\" Copyright (C) 2005, 2007  John Elliott
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH txt2psf 1 "21 June, 2008" "Version 1.0.8" "PSF Tools"
.\"
.\"------------------------------------------------------------------
.\"
.SH NAME
txt2psf - compile a PC Screen Font file from a textual description
.\"
.\"------------------------------------------------------------------
.\"
.SH SYNOPSIS
.PD 0
.B txt2psf
.RI [ --psf1 ]
.RI [ --psf2 ]
.RI [ INPUTFILE 
.RI [ OUTPUTFILE ]]
.P
.PD 1
.\"
.\"------------------------------------------------------------------
.\"
.SH DESCRIPTION
.LP 
txt2psf converts an ASCII source file to a font in the .PSF format.
.\"
.\"------------------------------------------------------------------
.\"
.SH INPUT FILE FORMAT
.LP
The source file is composed of a header, followed by a series of 
character definitions. The first line of the header must be "%PSF2"; it
is followed by these fields, each on a separate line, in any order:
.TP
.B "Version: nnn"
File format version; currently this must be 0.
.TP
.B "Flags: nnn"
This is 1 if the font should include a Unicode table, else 0.
.TP
.B "Length: nnn"
The number of characters in this font. If it is too low, the extra 
characters will be ignored. Too high, and the extras will be left blank.
.TP
.B "Width: nnn"
The width of a character cell.
.TP
.B "Height: nnn"
The height of a character cell.
.LP
After the header, character definitions follow. Each character definition
starts as a "%" on a line by itself. It is then followed by two fields, in
any order.
.TP
.B "Bitmap: bitmapdata"
The bitmap is composed of width*height cells, each of which is either a '-' 
(zero) or a '#' (black). Whitespace is ignored. The bitmap can be 
spread over several lines; the bitmaps created by psf2txt(1) are, for example.
.TP
.B "Unicode: unicodestring"
This describes the Unicode characters this bitmap represents. It is formed 
"[single];[sing]e];...[sequence];[sequence];..." where 
.I single 
is a single hexadecimal number, and
.I sequence 
is two or more hexadecimal numbers separated by + signs. 
So "[41];" is an example of a single character, and "[00B4+0065];" 
is an example of a sequence. If the bitmap doesn't represent
any Unicode characters this field may be blank or simply missing.
.\"
.\"------------------------------------------------------------------
.\"
.SH OPTIONS
.TP
.B --psf1
Force the resulting file to be in PSF1 format.
.TP
.B --psf2
Force the resulting file to be in PSF2 format.
.\"
.\"------------------------------------------------------------------
.\"
.SH BUGS
.LP
The file format is too rigid, and the parser (such as it is) a quick hack.
.LP
The "Length:" header should not be necessary; txt2psf should be able to
calculate the file size itself.
.\"
.\"------------------------------------------------------------------
.\"
.SH SEE ALSO
psf2txt(1)
.\"
.\"------------------------------------------------------------------
.\"
.SH AUTHOR
John Elliott <seasip.webmaster@gmail.com>.
.PP


psftools/NEWS
What's new in psftools-1.0.12:

	* fnt2psf: Changed the Windows 1.x font extract to seek to the
	 character bitmap rather than assuming it directly follows the
	 widths table. 



psftools/tools/txt2psf.c
/*
    psftools: Manipulate console fonts in the .PSF format
    Copyright (C) 2003  John Elliott

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
#include "cnvshell.h"
#include "psflib.h"

/* Convert an MDA character ROM to a .PSF font. 
 * 
 * We assume the MDA ROM format, with 256 characters, each one 8x14. The 
 * ninth column is synthesized.
 */

static char helpbuf[2048];
static int v1 = 0;
static int v2 = 0;
static PSF_FILE psf;

/* Program name */
char *cnv_progname = "TXT2PSF";

/* ddash = 1 if option started with a double-dash; else 0 */
/* Return NULL if OK, else error string */
char *cnv_set_option(int ddash, char *variable, char *value)
{
	if (!stricmp(variable, "psf1")) { v1 = 1; return NULL; }
	if (!stricmp(variable, "psf2")) { v2 = 1; return NULL; }

	if (strlen(variable) > 2000) variable[2000] = 0;
	sprintf(helpbuf, "Unknown option: %s\n", variable);
	return helpbuf;
}


/* Return help string */
char *cnv_help(void)
{
    sprintf(helpbuf, "Syntax: %s textfile psffile { options }\n\n", cnv_progname);
    strcat (helpbuf, "Options:\n\n"
		    "--psf1: Save in PSF1 format.\n"
		    "--psf2: Save in PSF2 format.\n");
            
    return helpbuf;
}

#define ATSTART  0
#define INHEADER 1
#define INCHAR   2
#define INBITMAP 3

char *cnv_execute(FILE *infile, FILE *outfile)
{	
	int rv, n, x = 0, y = 0;
	int state = 0;	
	char linebuf[2000];
	char unibuf[2000];
	char *c;
	int version = -1, flags = -1, length = -1, width = -1, height = -1;
	int nchar = 0;
	int line = 0, wb = 0, havebitmap = 0;
	psf_byte *charbits = NULL;

	psf_file_new(&psf);
	while (fgets(linebuf, sizeof(linebuf), infile))
	{
		++line;
/* If a long line, devour the rest of it */
		if (!strchr(linebuf, '\n'))
		{
			do
			{
				n = fgetc(infile);
			}
			while (n != EOF && n != '\n');	
		}
		c = strchr(linebuf, '\n'); if (c) *c = 0;
		c = strstr(linebuf, "//"); if (c) *c = 0;

		if (linebuf[0] == 0) continue;
		if (linebuf[0] == '%')
		{
			if (!strncmp(linebuf + 1, "PSF2", 4) && state == ATSTART)
			{
				state = INHEADER;
				continue;
			}
/* %STOP: Immediately stop parsing */
			else if (!strncmp(linebuf + 1, "STOP", 4))
			{
				break;	
			}	
			else if (state == INHEADER)
			{
				if (version == -1) return "Version not set";
				if (flags == -1)   return "Flags not set";
				if (width == -1)   return "Width not set";
				if (height == -1)  return "Height not set";
				if (length == -1)  return "Length not set";
				if (version > 0)   return "Versions greater than 0 are not supported"; 
				rv = psf_file_create(&psf, width, height,
					       length, (flags & 1));	
				if (!rv) 
				{
					charbits = malloc(psf.psf_charlen);
					if (!charbits) rv = PSF_E_NOMEM;
				}
				if (rv) return psf_error_string(rv);
				state = INCHAR;
				unibuf[0] = 0;
				wb = ((width + 7) / 8);
			}
			else if (state == INCHAR || state == INBITMAP)
			{
				if (version == -1) return "Version not set";
				if (nchar < psf.psf_length) memcpy
					(psf.psf_data + nchar * psf.psf_charlen,
					charbits, psf.psf_charlen);
				if (unibuf[0])
				{
					rv = psf_unicode_from_string(&psf, nchar, unibuf);
					if (rv) 
					{
						fprintf(stderr, "Line %d: Failed to decode %s\n", line, unibuf);
						return psf_error_string(rv);
					}
				}
				havebitmap = 0;
				state = INCHAR;
				nchar++;
				unibuf[0] = 0;
			}
			else 
			{
				fprintf(stderr, "Line %d: Unexpected %% line\n", line);
				return "Invalid input format";
			}	
		}
		else
		{
			c = strstr(linebuf, ": ");
			if (c)
			{
				*c = 0;
				c++;
				while (*c == ' ') c++;
				if (state == INHEADER)
				{
					if (!strcmp(linebuf, "Version")) 
						version = atol(c);
					else if (!strcmp(linebuf, "Flags"))
						flags = atol(c);
					else if (!strcmp(linebuf, "Width"))
						width = atol(c);
					else if (!strcmp(linebuf, "Height"))
						height = atol(c);
					else if (!strcmp(linebuf, "Length"))
						length = atol(c);
					else
					{
						fprintf(stderr, "Line %d: Unknown variable name '%s'\n", line, linebuf);
						return "Invalid input format";
					}
				}
				else if (state == INCHAR)
				{
					if (!strcmp(linebuf, "Unicode"))
						strcpy(unibuf, c);
					else if (!strcmp(linebuf, "Bitmap"))
					{
						strcpy(linebuf, c);
						state = INBITMAP;	
						havebitmap = 0;
						x = y = 0;
					} 
					else
					{
						fprintf(stderr, "Line %d: Unknown variable name '%s'\n", line, linebuf);
						return "Invalid input format";
					}
				}
				else	/* Not INCHAR and not INHEADER */
				{
					fprintf(stderr, "Line %d: Unknown variable name '%s'\n", line, linebuf);
					return "Invalid input format";
				}
				if (state != INBITMAP) continue;
			} /* end if (c) */
			if (state == INBITMAP)
			{
				psf_byte *dest, mask;
				c = linebuf;
				while (*c)
				{
					if (*c != '-' && *c != '#') 
					{
						c++;
						continue;
					}
					dest = charbits + (y * wb) + (x/8);
					mask = 0x80 >> (x & 7);
					if (*c == '#') *dest |= mask;
					else	       *dest &= ~mask;
					x++;
					if (x >= psf.psf_width) { x = 0; y++; }
					if (y >= psf.psf_height) 
					{
						state = INCHAR;
						havebitmap = 1;	
						break;
					}
					++c;
				}
				continue;
			}	
			fprintf(stderr, "Line %d: Not a Variable: Value pair\n", line);
			return "Invalid input format";
		}
	}
	if (state == INCHAR && havebitmap)
	{
		if (nchar < psf.psf_length) memcpy(psf.psf_data + nchar * 
				psf.psf_charlen, charbits, psf.psf_charlen);
		if (unibuf[0])
		{
			rv = psf_unicode_from_string(&psf, nchar, unibuf);
			if (rv) return psf_error_string(rv);
		}
	}
	if (v1) psf_force_v1(&psf);
	if (v2) psf_force_v2(&psf);
	rv = psf_file_write(&psf, outfile);
	psf_file_delete(&psf);	
	if (rv) return psf_error_string(rv);
	return NULL;
}

